### PR TITLE
#1508 - Support character-level external recommenders

### PR DESCRIPTION
**Pull request by reckart**
**What's in the PR**
- Adjusted suggestion extraction for the different anchoring modes
- Allow configuring external recommenders for all anchoring modes

**How to test manually**
* Try the usual recommenders on different levels of granularity
* Try using an external recommender on the character-level granularity

**Automatic testing**
* [ ] PR includes unit tests

**Documentation**
* [ ] PR updates documentation

### CODE REVIEW

**Comment by jcklie** [comment]
Works for me with a stupid stemming example:

image

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 1]
```diff
@@ -1187,34 +1189,86 @@ public Predictions computePredictions(User aUser, Project aProject,
...
+            case CHARACTERS: {
+                int[] offsets = { annotationFS.getBegin(), annotationFS.getEnd() };
+                TrimUtils.trim(annotationFS.getCAS().getDocumentText(), offsets);
+                begin = offsets[0];
+                end = offsets[1];
+                break;
+            }
+            case SINGLE_TOKEN: {
+                List<AnnotationFS> tokens = selectCovered(tokenType, annotationFS);
+                if (tokens.isEmpty()) {
```
Comment (jcklie) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: I would maybe add some logging here.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 2]
```diff
@@ -1187,34 +1189,86 @@ public Predictions computePredictions(User aUser, Project aProject,
...
+                    // remote service performs its own tokenization). We might be smart here by
+                    // looking for overlapping tokens instead of contained tokens.
+                    continue;
+                }
+                
+                AnnotationFS firstToken = tokens.get(0);
+                AnnotationFS lastToken = tokens.get(tokens.size() - 1);
+                
+                if (!firstToken.equals(lastToken)) {
+                    // We only want to accept single-token suggestions
```
Comment (jcklie) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: I would maybe add some logging here.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 3]
```diff
@@ -1187,34 +1189,86 @@ public Predictions computePredictions(User aUser, Project aProject,
...
+                
+                begin = firstToken.getBegin();
+                end = lastToken.getEnd();
+                break;
+            }
+            case TOKENS: {
+                List<AnnotationFS> tokens = selectCovered(tokenType, annotationFS);
+                if (tokens.isEmpty()) {
+                    // This can happen if a recommender uses different token boundaries (e.g. if a
+                    // remote service performs its own tokenization). We might be smart here by
```
Comment (jcklie) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: I would maybe add some logging here.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java [hunk 4]
```diff
@@ -1187,34 +1189,86 @@ public Predictions computePredictions(User aUser, Project aProject,
...
+                
+                begin = tokens.get(0).getBegin();
+                end = tokens.get(tokens.size() - 1).getEnd();
+                break;
+            }
+            case SENTENCES: {
+                List<AnnotationFS> sentences = selectCovered(sentenceType, annotationFS);
+                if (sentences.isEmpty()) {
+                    // This can happen if a recommender uses different token boundaries (e.g. if a
+                    // remote service performs its own tokenization). We might be smart here by
```
Comment (jcklie) [inline inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: I would maybe add some logging here.

File: inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java (hunk 2 above)
Comment (reckart) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: We could add debug logging, but I wouldn't add info logging.

Comment (jcklie) [reply inception-recommendation/src/main/java/de/tudarmstadt/ukp/inception/recommendation/service/RecommendationServiceImpl.java]: That would be good, I agree in that info would be too much.

**Comment by jcklie** [comment]
Also works for chinese segmentation:

image

**Comment by jcklie** [comment]
@reckart 22:51:44 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check (checkstyle-check) on project inception-imls-external: You have 1 Checkstyle violation. -> [Help 1]

**Pull request merged**